### ISSUE TITLE
Workflow for Python packages

### ISSUE THREAD

**Issue by ianlewis** [slsa-framework/slsa-github-generator#55]
Add a workflow for building python packages and generating SLSA provenance for them. This is analogous to https://github.com/slsa-framework/slsa-github-generator-go for Go projects, except for Python packages.

**Comment by ianlewis** [comment]
Here is a summary of the relevant file formats and tools to consider when building python packages. The [Python Packaging User Guide](https://packaging.python.org/en/latest/) provides a good summary of the available tools and methods.

Project definition files:
- [pyproject.toml](https://python-poetry.org/docs/pyproject/) - this is the de-facto project definition and is defined in [PEP-518](https://peps.python.org/pep-0518/).
- [setup.cfg](https://packaging.python.org/en/latest/tutorials/packaging-projects/#configuring-metadata) - pre-PEP-518 configuration for project used by setuptools.

# Tools to build packages
These tools and packages support building packages based on either `pyproject.toml` or `setup.cfg`. For the most part if the project uses `pyproject.toml` the build tool shouldn't matter too much but some of the tools seem to have plugins and extensions which will be build tool specific.

- [build](https://pypa-build.readthedocs.io/en/stable/index.html) seems to be the de-facto build tool. It was built as a result of [PEP-517](https://peps.python.org/pep-0517/) and [PEP-518](https://peps.python.org/pep-0518/) which aimed to replace distutils and setuptools.
- [Poetry](https://python-poetry.org/) is a fairly popular project for Python packages analogous to [goreleaser](https://goreleaser.com/). It uses a toml config file which we could pick up on and use. Or we could consider using our own config file that is compatible. It has it's own [pyproject.yaml extensions](https://python-poetry.org/docs/pyproject/).
- [hatch](https://github.com/pypa/hatch) is a python package build tool. Unsure of how widely adopted it is. **Need to investigate plugins and extensions.**
- [flit](https://flit.pypa.io/en/latest/) is a python package build tool. Unsure of how widely adopted it is. **Need to investigate plugins and extensions.**
- [setuptools](https://setuptools.pypa.io/en/latest/) uses `setup.cfg`. Its use is discouraged in favor if PEP-518 compatible tools but IIUC still has significant adoption.

# Tools to publish packages
Support for publishing provenance could maybe be added to these tools after pypi has support?

- [twine](https://pypi.org/project/twine/) is a tool to upload a pre-built package to pypi.
- [hatch](https://github.com/pypa/hatch) supports uploading to pypi.
- [flit](https://flit.pypa.io/en/latest/) supports uploading to pypi.

# Tools used to install packages

- [pip](https://pip.pypa.io/en/stable/) is the de-facto python package installer. It's very widely used and is ubiquitous enough that we can probably ignore other installers of pypi packages.

# Scientific community
The scientific community seems to use other pypi/pip incompatible package managers fairly frequently due to needing access to specialized libraries and tools.
- [conda](https://conda.io/en/latest/index.html) - mostly used in the scientific community. AFAICT the packages are incompatible with pip packages and are served on a totally different infrastructure from pypi.
- [spack](https://spack.readthedocs.io/en/latest/index.html) - used mostly by the scientific community(?)
- [hashdist](https://hashdist.readthedocs.io/en/latest/) - used mostly by the scientific community(?)

**Comment by ianlewis** [comment]
@di I summarized some info on python package builders above with the goal of better understanding we could put together a builder that will be able to build python packages and generate SLSA provenance for it.

As I understand it, if folks wrote pypackage.yaml in such a way as to set `requires`, `build-backend`, and `build-path` properly then theoretically any package build using pypackage.yaml could be bootstrapped by the `build` tool.

However, I don't think most packages are set up this way and, without modification, will require us to know the build tool up front in order to build the package properly. That or we force them to update their package to allow building with the `build` tool. And many folks still use setuptools.

Any thoughts or comments? Anything I got wrong or am misunderstanding?

**Comment by di** [comment]
> Any thoughts or comments? Anything I got wrong or am misunderstanding?

I think this is fairly accurate. I'd say you should think of `build` as the canonical generic build tool, and everything else (flit, setuptools, etc) as PEP-517 build backends (even though some can also be used as build tools).  

> As I understand it, if folks wrote pypackage.yaml in such a way as to set requires, build-backend, and build-path properly then theoretically any package build using pypackage.yaml could be bootstrapped by the build tool.

This is correct (although it's `pypackage.toml` 😉).

> However, I don't think most packages are set up this way and, without modification, will require us to know the build tool up front in order to build the package properly. That or we force them to update their package to allow building with the build tool. And many folks still use setuptools.

I think it's really our only option. We can do some things to guess the build backend that needs to be present, but it's not guaranteed to be correct, or capture all bulid-time dependencies. Instead, `pyproject.toml` is the standardized way to do this. I think it would be reasonable to say we only support building projects that specify `pyproject.toml`.

From my perspective, we're seeing a lot of projects moving to this (and a lot of tools moving their configuration into this file).

**Comment by ianlewis** [comment]
@di Thanks. I think `build` is definitely the thing we should do first. I think you're right that a decent number of folks use `pyproject.toml` but I didn't see many in my searches that set the backends properly. They looked like they expected you to install and/or run the build backend directly.

That said, it should be pretty easy for folks to update their `pyproject.toml` to allow us to just run `python -m build` on their project and have it work, so maybe that will cover ~80% of projects. Folks still using setuptools only would have issues though but maybe we can support that at some point if it's as issue.

Scientific community projects also worry me because lots use `conda`, but I'm ok with not supporting them, at least initially.

**Comment by di** [comment]
> I think build is definitely the thing we should do first. I think you're right that a decent number of folks use pyproject.toml but I didn't see many in my searches that set the backends properly. They looked like they expected you to install and/or run the build backend directly.

Note that build uses a fallback backend if none is specified, so that might be the case: https://python-build.readthedocs.io/en/stable/#fallback-backend

**Comment by ianlewis** [comment]
> Note that build uses a fallback backend if none is specified, so that might be the case: https://python-build.readthedocs.io/en/stable/#fallback-backend

Ah, ok. Noted. Though I did see a lot of `pyproject.toml` files with `[tool.foo]` sections but without any `requires` or backend set so I expect those not to work.

**Comment by di** [comment]
Yeah, these would be the projects that `build` would use the fallback backend for. The entire `[build-system]` section would be missing.

**Comment by ianlewis** [comment]
> Yeah, these would be the projects that `build` would use the fallback backend for. The entire `[build-system]` section would be missing.

If, for example, it's exclusively `[tool.poetry.x]` sections in `pypackage.toml`, I wouldn't expect `python -m build` to work because `build` would fallback to setuptools. For example: https://github.com/razy69/poetry-tox-template/blob/208cbc3f7d972a0c7862c2f54c9dfe3fe5b74e54/pyproject.toml

If there was a separate and analogous `setup.cfg` or `setup.py` it would work though I guess. 

Am I understanding it right? Or would `build` figure it out somehow?

**Comment by ianlewis** [comment]
We may also want to consider not having a builder in this repo but one that is better endorsed by the python commmunity. e.g. via https://github.com/pypa

**Comment by reneleonhardt** [comment]
Is there an update for building Python wheels with cibuildwheel?
